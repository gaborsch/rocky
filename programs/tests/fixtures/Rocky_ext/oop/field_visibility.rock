a cat looks like nothing
  the food is nowhere
  a cat takes the milk
    let the food be the milk
  (eom - constructor)
  the food takes something
    give back the food
  (eom)
  Make Noise takes something
    say the noise
  (eom)
(eoc)

a tiger looks like a cat
  a tiger takes the prey
     let the food be the prey
  (eom - constructor)
  the prey takes something
    give back the food
  (eom)
  the noise is nothing
  Set Noise takes a roar
    let the noise be a roar
  (eom)
(eoc)

say "Field access"
Kitty wants to be a cat taking "milk"
say the food by Kitty taking nothing (has access to own field)

Tom wants to be a tiger taking "Jerry"
say the food by Tom taking nothing (inherited method has access to local variable)
say the prey by Tom taking nothing (method has access to parent class fields)

say "Abstract field access"

(Make Noise by Kitty taking nothing) (runtime error - field 'the noise' is not defined in parent class)

Set Noise for Tom taking "Roarrrr!"
(Make Noise near Tom taking nothing) (runtime error - parent method has NO access to child-defined fields)


