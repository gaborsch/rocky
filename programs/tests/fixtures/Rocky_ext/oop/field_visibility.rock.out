Field access
milk
Jerry
Jerry
Abstract field access
